SQL/Release_patches/20.3_To_21.0_upgrade.sql

```diff
@@ -281,7 +281,7 @@ FROM Project JOIN subproject
 WHERE Project.Name='loris';
 
 -- if the loris project was added, set all candidates to that default project
-UPDATE candidate SET ProjectID=(SELECT ProjectID FROM Project WHERE Name='loris') WHERE ProjectID IS NULL AND Entity_type='Human';
+UPDATE candidate SET ProjectID=(SELECT ProjectID FROM Project WHERE Name='loris') WHERE ProjectID IS NULL;
```

Comment:
The default project named "loris" is added in this patch for 21.0.. it won't have existed in 20.3 and is only created above for projects that didn't previously use projects.

I'm not really sure what the right way to fix this is. Options are:
1. Always create I project named "loris" rather than only if there are no current projects
2. Make the default MIN(Project.ID) (or MAX) instead of Name="loris"
3. Write a script instead of an SQL patch to handle both cases before running the schema upgrade patch.

3 is probably the most robust, but I'm not sure how much extra work we want to be adding to writing/testing at this point in the release process

Reply:
I don't think there is any harm in creating a loris project for everyone.

Reply:
Not sure i agree. Last time we discussed we said the "loris" default would be there unless its replaced by the study managers. The patch does exactly that, if there are no projects it adds one. If there already are projects it assumes the study managers will use one of these for the different candidates sessions and fantoms since they determined these projects themselves.

I feel like the current state is the closest to what we had previously discussed. We can add release statement describing that or a php script to just check the state of the database but i dont see it more complexe than that

Reply:
The problem isn't that the patch does that for projects that don't have any projects set up, the problem is that the update statement trying to make sure there's no null ProjectID will fail (or at least, not update any rows) for projects that *do* already have some projects set up (unless they happen to have one named "loris")

Reply:
@driusan in the current state of the upgrade patch no errors would be generated regardless of the state of your project table.

the following behaviour is to be expected.

 - if the project table is empty
  1. a new project "loris will be added"
  2. all candidates will be assigned to the new project

- if the project table already has projects:
 1. no new project is added
 2. candidates with no project affiliation will remain as they are (makes logical sens since if the projects table was populated, the useprojects was probably on and candidates SHOULD already belong to projects)

what I can do is add a php script that runs after the SQL to remind projects to add a project ID to their candidates with no projects

Reply:
In that case I don't think we need a PHP script, we can just mention in the release notes that moving forward, all candidates should have projects and they should verify the state of their database